gnomad/utils/constraint.py

```diff
@@ -0,0 +1,177 @@
+
+    grouping = hl.struct(context=ht.context, ref=ht.ref, alt=ht.alt)
+    if not omit_methylation:
+        grouping = grouping.annotate(methylation_level=ht.methylation_level)
+    for group in additional_grouping:
+        grouping = grouping.annotate(**{group: ht[group]})
+
+    agg = {
+        "variant_count": hl.agg.count_where(freq_expr[0].AF <= max_af)
+        if max_af
+        else hl.agg.count()
+    }
+    if count_singletons:
+        agg["singleton_count"] = hl.agg.count_where(singleton_expr)
+
+    for pop in count_downsamplings:
+        agg[f"downsampling_counts_{pop}"] = downsampling_counts_expr(
+            freq_expr, freq_meta_expr, pop, max_af=max_af
+        )
+        if count_singletons:
+            agg[f"singleton_downsampling_counts_{pop}"] = downsampling_counts_expr(
+                freq_expr, freq_meta_expr, pop, singleton=True
+            )
+
+    if use_table_group_by:
+        return ht.group_by(**grouping).partition_hint(partition_hint).aggregate(**agg)
+    else:
+        return ht.aggregate(
+            hl.struct(**{field: hl.agg.group_by(grouping, agg[field]) for field in agg})
+        )
+
+
+def downsampling_counts_expr(
+    freq_expr: hl.expr.ArrayExpression,
+    freq_meta_expr: hl.expr.ArrayExpression,
+    pop: str = "global",
+    variant_quality: str = "adj",
+    singleton: bool = False,
+    max_af: Optional[float] = None,
+) -> hl.expr.ArrayExpression:
+    """
+    Count variants in `freq_expr` of downsamplings with specified criteria in their meta data.
+
+    The meta data (`freq_meta_expr`) of each downsampling should have 'group',
+    'pop', and 'downsampling' keys, where 'group' key is 'variant_quality' and 'pop' key is 'pop'.
+
+    :param freq_expr: ArrayExpression of Structs with with 'AC' and 'AF' annotations.
+    :param freq_meta_expr: ArrayExpression of meta dictionaries corresponding to freq_expr.
+    :param pop: Population. Defaults to 'global'.
+    :param variant_quality: Variant quality for "group" key in `freq_meta_expr`. Defaults to 'adj'.
+    :param singleton: Whether to sum only alleles that are singletons. Defaults to False.
+    :param max_af: Maximum variant allele frequency to keep. By default no allele frequency cutoff is applied.
+    :return: Variant counts in downsamplings for specified population.
+    """
+    # get indices of dictionaries in meta dictionaries that only have the "downsampling" key with required "group" and "pop" keys
+    indices = hl.enumerate(freq_meta_expr).filter(
+        lambda f: (f[1].size() == 3)
+        & (f[1].get("group") == variant_quality)
+        & (f[1].get("pop") == pop)
+        & f[1].contains("downsampling")
+    )
+    # get an array of indices sorted by "downsampling" key
+    sorted_indices = hl.sorted(indices, key=lambda f: hl.int(f[1]["downsampling"])).map(
+        lambda x: x[0]
+    )
+
+    def _get_criteria(i: hl.expr.Int32Expression) -> hl.expr.ArrayNumericExpression:
```

Comment:
Just double check that the return type should be `ArrayNumericExpression` instead of `Int32Expression`.